File: examples/canonical-aws-lambda/s3-trigger-image-moderation/go/src/handler.go
package handler

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

type Event struct {
	Records []Record
}

type Record struct {
	S3 struct {
		Bucket struct {
			Name string
		}
		Object struct {
			Key       string
			VersionID string
			Sequencer string
		}
	}
}

type Clients interface {
	ClaimObjectVersion(ctx context.Context, key string) (bool, error)
	DetectLabels(ctx context.Context, bucket, key string) ([]string, error)
	DetectModerationLabels(ctx context.Context, bucket, key string) ([]string, error)
	RecordDecision(ctx context.Context, key, decision string, labels []string) error
	PublishFlagged(ctx context.Context, detail map[string]string) error
}

type Result struct {
	Processed  int
	Duplicates int
	Decision   string
}

func Handle(ctx context.Context, event Event, clients Clients) (Result, error) {
	started := time.Now()
	result := Result{}
	for _, record := range event.Records {
		version := record.S3.Object.VersionID
		if version == "" {
			version = record.S3.Object.Sequencer
		}
		if version == "" {
			version = "unversioned"
		}
		dedupeKey := fmt.Sprintf("%s:%s:%s", record.S3.Bucket.Name, record.S3.Object.Key, version)
		claimed, err := clients.ClaimObjectVersion(ctx, dedupeKey)
		if err != nil {
			return result, err
		}
		if !claimed {
			result.Duplicates++
			logShape(started, dedupeKey, "DROP", "duplicate object version")
			continue
		}
		labels, err := clients.DetectLabels(ctx, record.S3.Bucket.Name, record.S3.Object.Key)
		if err != nil {
			return result, err
		}
		moderation, err := clients.DetectModerationLabels(ctx, record.S3.Bucket.Name, record.S3.Object.Key)
		if err != nil {
			return result, err
		}
		decision := "ALLOW"
		if len(moderation) > 0 {
			decision = "FLAG"
			if err := clients.PublishFlagged(ctx, map[string]string{"bucket": record.S3.Bucket.Name, "key": record.S3.Object.Key, "version": version, "decision": decision}); err != nil {
				return result, err
			}
		}
		if err := clients.RecordDecision(ctx, dedupeKey, decision, labels); err != nil {
			return result, err
		}
		result.Processed++
		result.Decision = decision
		logShape(started, dedupeKey, decision, "moderation complete")
	}
	return result, nil
}

func logShape(started time.Time, dedupeKey, decision, msg string) {
	entry := map[string]any{
		"timestamp":     time.Now().UTC().Format(time.RFC3339),
		"level":         "INFO",
		"msg":           msg,
		"trace_id":      "local-trace",
		"request_id":    dedupeKey,
		"correlation_id": dedupeKey,
		"provider":      "aws",
		"runtime_tier":  "function",
		"function_name": "accb-dev-s3mod-handler",
		"dedupe_key":    dedupeKey,
		"decision":      decision,
		"latency_ms":    time.Since(started).Milliseconds(),
	}
	payload, _ := json.Marshal(entry)
	fmt.Println(string(payload))
}

File: examples/canonical-aws-lambda/s3-trigger-image-moderation/typescript/src/handler.js
"use strict";

async function handler(event, clients) {
  const started = Date.now();
  const record = event.Records[0];
  const bucket = record.s3.bucket.name;
  const object = record.s3.object;
  const version = object.versionId || object.sequencer || "unversioned";
  const dedupeKey = `${bucket}:${object.key}:${version}`;
  const requestId = event.requestContext?.requestId || "local-request";
  const active = clients || awsClients();
  const claimed = await active.claimObjectVersion(dedupeKey, { bucket, key: object.key, version });
  if (!claimed) {
    log(started, requestId, dedupeKey, dedupeKey, "DROP", "duplicate object version");
    return { ok: true, duplicate: true, decision: "DROP" };
  }
  const labels = await active.detectLabels(bucket, object.key);
  const moderation = await active.detectModerationLabels(bucket, object.key);
  const decision = moderation.length > 0 ? "FLAG" : "ALLOW";
  await active.recordDecision(dedupeKey, decision, labels);
  if (decision === "FLAG") {
    await active.publishFlagged({ bucket, key: object.key, version, decision });
  }
  log(started, requestId, dedupeKey, dedupeKey, decision, "moderation complete");
  return { ok: true, duplicate: false, decision };
}

function awsClients() {
  throw new Error("AWS clients are bound in the deployment package; tests inject clients.");
}

function log(started, requestId, correlationId, dedupeKey, decision, msg) {
  console.log(JSON.stringify({
    timestamp: new Date().toISOString(),
    level: "INFO",
    msg,
    trace_id: process.env._X_AMZN_TRACE_ID || "local-trace",
    request_id: requestId,
    correlation_id: correlationId,
    provider: "aws",
    runtime_tier: "function",
    function_name: process.env.AWS_LAMBDA_FUNCTION_NAME || "accb-dev-s3mod-handler",
    dedupe_key: dedupeKey,
    decision,
    latency_ms: Date.now() - started
  }));
}

module.exports = { handler };
